Fix sample weight handling in scoring _log_reg_scoring_path

snath-xoc opened the pull request:
#### Reference Issues/PRs
Fixes #29416 


#### What does this implement/fix? Explain your changes.

Added sample weighting for test set into default calculation of scores within _log_reg_scoring_path

TO DO:

- [ ] so far works with max_iter 10_000 and tol 1e-8, expected to work with tol 1e-12 but this fails on lbfgs solver
- [x] Switch changelog to v1.6.rst

Commit:
add sample weight to default scoring _log_reg_scoring_path

ogrisel commented:
Thanks for the PR. Could you please add a non regression test based on the reproducer provided in #29416 and make it run on all solvers with a small value of `tol` such as `tol=1e-12`? It's possible that strict equivalence might be hard to reach for stochastic solvers such as SAG/SAGA but I am not sure.

snath-xoc committed:
merged test for sample weight into test_logistic_regression_sample_weights

snath-xoc commented:
Note: in the test_logistic_regression_sample_weight sag and saga are left out as they systematically fail

ogrisel commented:
@snath-xoc could you please push a commit that triggers running the updated test in this PR for all admissible values of `global_random_seed`? Instructions are available in this information issue: #28959.

ogrisel commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -775,44 +776,54 @@ def test_logistic_regressioncv_class_weights(weight, class_weight, global_random
+            clf_sw_repeated.fit(X_repeated, y_repeated)
+            if LR is LogisticRegressionCV:
+                assert_allclose(clf_sw_weighted.scores_[1], clf_sw_repeated.scores_[1])
+            assert_allclose(clf_sw_weighted.coef_, clf_sw_repeated.coef_, rtol=1e-4)
```
`rtol=1e-4` seems like a large tolerance level for comparing coefficients...

Maybe the tolerance of the solvers is too high, and the two models do not converge properly enough to be able to test 

I did some local experiments, and I think we could always use:

```python
    max_iter = 10_000
    tol = 1e-12
```

for all estimators in this test and the check for `assert_allclose` on coefs with `rtol=1e-5` and `atol=1e-8`, including with `sag` and `saga` solvers.

ogrisel reviewed (commented):
Here is some feedback to help move this PR forward. Please don't forget to add an entry to the changelog.

ogrisel commented:
I find this test too long to follow. You might want to decouple the part that test equivalence between integer sample weights and their repeated counter part from the part of the test that checks equivalence between class weights and sample weights into two independent tests.

snath-xoc committed:
modified test_logistic [all random seeds]

snath-xoc committed:
modified changelog [all random seeds]

snath-xoc committed:
increase max iter in l1l2 liblinear tests [all random seeds]

snath-xoc committed:
fixed indent error [all random seeds]

ogrisel commented on doc/whats_new/v1.5.rst:
```diff
@@ -18,6 +18,20 @@ For a short description of the main highlights of the release, please refer to
+Changelog
+---------
+
+:mod:`sklearn.linear_model`
+..........................
```
```suggestion
:mod:`sklearn.linear_model`
...........................
```

This is causing the doc CI to fail with:

```
/home/circleci/project/doc/whats_new/v1.5.rst:44: WARNING: Title underline too short.
```

ogrisel commented on doc/whats_new/v1.5.rst:
```diff
@@ -18,6 +18,20 @@ For a short description of the main highlights of the release, please refer to
+..........................
+
+-|Fix| Fixed sample weight handling for calculation of test scores under  _log_reg_scoring_path
+:pr:`29419` by :user:`Shruti Nath <snath-xoc>`
```
Please only mention public API symbols in the changelog.

Also. Scikit-learn 1.5.1 is already released. The changelog entry should be moved to `v1.6.rst`.

ogrisel replied:
Here the most directly public symbol impacted by this fix is `LogisticRegressionCV`.

ogrisel commented:
To actually trigger `[all random seeds]` tests you need to include the list of test function names after the `[all random seeds]` flag in the commit message as explained in #28959.

snath-xoc committed:
[all random seeds]
test_logistic_regression_sample_weights

snath-xoc committed:
[all random seeds]
test_logistic_regression_sample_weights

snath-xoc committed:
fix tol for sag and saga [all random seeds]
test_logistic_regression_sample_weights

snath-xoc committed:
fix max_iter for sag and saga [all random seeds]
test_logistic_regression_sample_weights

snath-xoc committed:
Merge branch 'main' into logisticregressioncv_sample_weight

snath-xoc committed:
[all random seeds]
test_logistic_regression_sample_weights

snath-xoc committed:
[all random seeds]
test_sample_and_class_weight_equivalence_liblinear
test_logistic_regression_sample_weights
test_logistic_regression_solver_class_weights

snath-xoc committed:
[all random seeds]
test_sample_and_class_weight_equivalence_liblinear
test_logistic_regression_sample_weights
test_logistic_regression_solver_class_weights

snath-xoc committed:
[all random seeds]
test_sample_and_class_weight_equivalence_liblinear
test_logistic_regression_sample_weights
test_logistic_regression_solver_class_weights

snath-xoc committed:
[all random seeds]
test_sample_and_class_weight_equivalence_liblinear
test_logistic_regression_sample_weights
test_logistic_regression_solver_class_weights

OmarManzoor commented on doc/whats_new/v1.5.rst:
```diff
@@ -65,7 +65,7 @@ Changelog
 
 :mod:`sklearn.metrics`
-......................
+.......................
```
I don't think this file needs to be changed.

ogrisel replied:
Indeed, let's remove this change:

```suggestion
......................
```

OmarManzoor commented on doc/whats_new/v1.6.rst:
```diff
@@ -227,7 +227,9 @@ Changelog
 
 :mod:`sklearn.linear_model`
-...........................
+............................
+-|Fix| :class:`linear_model.LogisticRegressionCV` Fixed sample weight handling for calculation
+  of test scores under  _log_reg_scoring_path :pr:`29419` by :user:`Shruti Nath <snath-xoc>`.
```
```suggestion
...........................
- |Fix| :class:`linear_model.LogisticRegressionCV` corrects sample weight handling
  for the calculation of test scores.
  :pr:`29419` by :user:`Shruti Nath <snath-xoc>`.
```

ogrisel replied:
I edited Omar's suggestion to also insert a blank line after the section header to be consistent with the style of the rest of the file.

OmarManzoor reviewed (approved):
Otherwise LGTM. Thanks @snath-xoc

ogrisel commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -775,86 +776,164 @@ def test_logistic_regressioncv_class_weights(weight, class_weight, global_random
+    elif problem == "cv":
+        LR = LogisticRegressionCV
+        # We weight the first fold 2 times more.
+        groups_weighted = np.r_[
+            np.full(n_samples_per_cv_group, 0),
+            np.full(n_samples_per_cv_group, 1),
+            np.full(n_samples_per_cv_group, 2),
+        ]
```
I think `np.concatenate` is easier to understand (more explicit name).

```suggestion
        groups_weighted = np.concatenate([
            np.full(n_samples_per_cv_group, 0),
            np.full(n_samples_per_cv_group, 1),
            np.full(n_samples_per_cv_group, 2),
        ])
```

ogrisel commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -775,86 +776,164 @@ def test_logistic_regressioncv_class_weights(weight, class_weight, global_random
+    clf_sw_repeated = LR(solver=solver, **kw_repeated)
+
+    if solver == "lbfgs":
+        # lbfgs has convergence issues on the data
```
```suggestion
        # lbfgs has convergence issues on the data but this should not impact
        # the quality of the results.
```

ogrisel commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -775,86 +776,164 @@ def test_logistic_regressioncv_class_weights(weight, class_weight, global_random
+def test_logistic_regression_solver_class_weights(solver, global_random_seed):
+    # Test that passing class_weight as [1,2] is the same as
+    # passing class weight = [1,1] but adjusting sample weights
+    # to be 2 for all instances of class 2
```
class 1 is the second class (class 0 is the first class).

```suggestion
    # to be 2 for all instances of class 1.
```

ogrisel commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -775,86 +776,164 @@ def test_logistic_regressioncv_class_weights(weight, class_weight, global_random
+    "solver", ("lbfgs", "newton-cg", "newton-cholesky", "sag", "saga")
+)
+def test_logistic_regression_solver_class_weights(solver, global_random_seed):
+    # Test that passing class_weight as [1,2] is the same as
```
Style nitpick:

```suggestion
    # Test that passing class_weight as [1, 2] is the same as
```

ogrisel reviewed (approved):
Some more minor feedback on top of Omar's but otherwise, LGTM!

ogrisel commented:
Note for later: the `LogisticRegressionCV` class has the following tag to skip the common test:

```python
    def _more_tags(self):
        return {
            "_xfail_checks": {
                "check_sample_weights_invariance": (
                    "zero sample_weight is not equivalent to removing samples"
                ),
            }
        }
```

However `check_sample_weights_invariance` currently does not work well with `kind="zeros"` in the presence of a `cv` constructor parameter. This common test should be updated to drop samples without moving the CV boundaries for the remaining training points.

This should better be done in a dedicated PR though.

snath-xoc committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>
Co-authored-by: Omar Salman <omar.salman2007@gmail.com>

ogrisel committed:
Fix linting problem

ogrisel commented:
Merged! Thanks very much @snath-xoc!
